--- a/hw/xfree86/common/compiler.h
+++ b/hw/xfree86/common/compiler.h
@@ -117,7 +117,7 @@ extern int ffs(unsigned long);
 
 #  if !defined(__arm__)
 #   if !defined(__sparc__) && !defined(__sparc) && !defined(__arm32__) \
-      && !(defined(__alpha__) && defined(linux)) \
+      && !(defined(__alpha__) && !(defined(__avr32__)) && defined(linux)) \
       && !(defined(__ia64__) && defined(linux)) \
 
 extern void outb(unsigned short, unsigned char);
@@ -127,7 +127,7 @@ extern unsigned int inb(unsigned short);
 extern unsigned int inw(unsigned short);
 extern unsigned int inl(unsigned short);
 
-#   else /* __sparc__,  __arm32__, __alpha__*/
+#   else /* __sparc__,  __arm32__, __alpha__, __avr32__ */
 
 extern void outb(unsigned long, unsigned char);
 extern void outw(unsigned long, unsigned short);
@@ -136,7 +136,7 @@ extern unsigned int inb(unsigned long);
 extern unsigned int inw(unsigned long);
 extern unsigned int inl(unsigned long);
 
-#   endif /* __sparc__,  __arm32__, __alpha__ */
+#   endif /* __sparc__,  __arm32__, __alpha__, __avr32__ */
 #  endif /* __arm__ */
 
 extern unsigned long ldq_u(unsigned long *);
@@ -1302,7 +1302,7 @@ inl(unsigned short port)
 #    define mem_barrier()	eieio()
 #    define write_mem_barrier()	eieio()
 
-#elif defined(__arm__) && defined(__linux__)
+#elif (defined(__arm__) || defined(__avr32__)) && defined(__linux__)
 
 #define ldq_u(p)	(*((unsigned long  *)(p)))
 #define ldl_u(p)	(*((unsigned int   *)(p)))
@@ -1341,6 +1341,7 @@ xf_outl(unsigned short port, unsigned in
 #define outw xf_outw
 #define outl xf_outl
 
+#if defined(__arm__)
 #define arm_flush_cache(addr)						\
 do {									\
   register unsigned long _beg __asm ("a1") = (unsigned long) (addr);	\
@@ -1350,6 +1351,7 @@ do {									\
     : "=r" (_beg)							\
     : "0" (_beg), "r" (_end), "r" (_flg));				\
 } while (0)
+#endif /* __arm__ */
 
 #   else /* ix86 */
 
@@ -1363,7 +1365,7 @@ do {									\
 #    define write_mem_barrier()   /* NOP */
 
 #    if !defined(__SUNPRO_C)
-#    if !defined(FAKEIT) && !defined(__mc68000__) && !defined(__arm__) && !defined(__sh__) && !defined(__hppa__) && !defined(__s390__)
+#    if !defined(FAKEIT) && !defined(__mc68000__) && !defined(__arm__) && !defined(__avr32__) && !defined(__sh__) && !defined(__hppa__) && !defined(__s390__)
 #     ifdef GCCUSESGAS
 
 /*
@@ -1472,7 +1474,7 @@ inl(unsigned short port)
 
 #     endif /* GCCUSESGAS */
 
-#    else /* !defined(FAKEIT) && !defined(__mc68000__)  && !defined(__arm__) && !defined(__sh__) && !defined(__hppa__)*/
+#    else /* !defined(FAKEIT) && !defined(__mc68000__)  && !defined(__arm__) && !defined(__avr32__) && !defined(__sh__) && !defined(__hppa__)*/
 
 static __inline__ void
 outb(unsigned short port, unsigned char val)

